File: 4week/1task/Basic/List/List.cpp
#include <iostream>
#include <cstdlib>
#include "List.h"

List::List(int capacity, double multiplier) : capacity(capacity), current(0), multiplier(multiplier) {
    this->array = (int*)malloc(capacity*sizeof(int));
    
    if ( this->array == NULL ) {
        throw OutOfMemoryException();
    }
}

List::~List() {
    free(this->array);
}

int List::size() const {
    return this->current;
}

int List::max_size() const {
    return this->capacity;
}

void List::erase(int index) {
    if (index >= this->current || index < 0 || index > this->current) {
        throw OutOfMemoryException();
    }
    if ( this->current == 0 ) {
        throw ZeroLenException();
    }
    
    for ( int i = index; i < this->current; i++ ) {
        this->array[i] = this->array[i+1];
    }
    this->current -= 1;
}

void List::insert(int value, int index) {
    if (index < 0 || index > this->current) {
        throw OutOfMemoryException();
    } else {
        int newCurrent = this->current + 1;
        
        if ( newCurrent > this->capacity ) {
            int newCapacity = this->capacity * this->multiplier;
            int* newArray = (int*)realloc(array, newCapacity*sizeof(int));
                if ( newArray == NULL ) {
                    throw OutOfMemoryException();
                }
            this->capacity = newCapacity;
            this->array = newArray;
        }
        this->current = newCurrent;
        int i = newCurrent;
        
        for ( ; i > index; i-- ) {
            this->array[i] = this->array[i-1];
        }
        this->array[index] = value;
    }
}

int List::find(int value) const {
    int index = 0;
    for ( ; index < this->capacity ; index++) {
        if (this->array[index] == value) {
            return index;
        }
    }
    return -1;
}

void List::push_back(int value) {
    int newCurrent = this->current + 1;
    
    if ( newCurrent > this->capacity ) {
        int newCapacity = this->capacity * this->multiplier;
        int* newArray = (int*)realloc(array, newCapacity*sizeof(int));
        
        if ( newArray == NULL ) {
            throw OutOfMemoryException();
        }
        
        this->capacity = newCapacity;
        this->array = newArray;
    }
    this->array[current] = value;
    this->current = newCurrent;
}

int List::pop_back() {
    if ( this->current == 0 ) {
        throw ZeroLenException();
    }
    this->current -= 1;
    return this->array[current];
}

void List::sort() {
    for ( int i = 1; i < this->current; i++ ) {
        int j = i;
        int temp = this->array[i];
        
        for ( int prev = j - 1; j > 0 && temp < this->array[prev]; j--, prev-- ) {
            this->array[j] = this->array[prev];
        }
        this->array[j] = temp;
    }
}

int List::operator[](int index) const {
    return this->array[index];
}

bool List::operator==(const List& other) const {
    return this->capacity == other.capacity;
}

bool List::operator!=(const List& other) const {
    return this->capacity != other.capacity;
}

std::ostream& operator<<(std::ostream& out, const List& list) {
   int last = list.size() - 1;

   for ( int i = 0; i < last; i++ ) {
       out << list[i] << ' ';
   }
   out << list[last];
   return out;
}

File: 4week/3task/List/List.java
public class List {
    private int[] array;
    private int capacity;
    private double multiplier;
    private int current;

    public List(int capacity, double multiplier) throws OutOfMemoryException {
        this.capacity = capacity;
        this.multiplier = multiplier;
        this.current = 0;

        array = new int[capacity];

        if ( array == null ) {
            throw new OutOfMemoryException();
        }
    }

    public int size() {
        return this.current;
    }

    public int max_size() {
        return this.capacity;
    }

    public void erase(int index) throws ZeroLenException {
        if ( index >= 0 ) {
            for ( int i = index; i < current; i++ ) {
            array[i] = array[i+1];
            }
            pop_back();
        }
    }

    public void insert(int value, int index) throws OutOfMemoryException {
        if ( index >= 0 && index <= current ) {
            push_back(0);
            for ( int i = current - 1; i > index; i-- ) {
                array[i] = array[i-1];
            }
            array[index] = value;
        }
    }

    public int pop_back() throws ZeroLenException {
        if ( current <= 0 ) {
            throw new ZeroLenException();
        }
        current -= 1;
        return array[current];
    }

    public int find(int value) {
        for ( int i = 0; i < current; i++ ) {
            if ( array[i] == value ) {
                return i;
            }
        }
        return -1;
    }

    public void push_back(int value) throws OutOfMemoryException {
        int newCurrent = current + 1;

        if ( newCurrent > capacity ) {
            int newCapacity = (int)(capacity * multiplier);
            int[] newArray = new int[newCapacity];

            if ( newArray == null) {
                throw new OutOfMemoryException();
            }

            for ( int i = 0; i < capacity; i++ ) {
                newArray[i] = array[i];
            }
            capacity = newCapacity;
            array = newArray;
            newArray = null;
        }
        array[current] = value;
        current = newCurrent;
    }

    public void sort() {
        int buffer;
        boolean isSorted = false;

        for ( int i = current - 1; i > 0 && isSorted == false; i-- ) {
            isSorted = true;
            for ( int j = 0; j < i; j++ ) {
                if ( array[j] > array[j+1] ) {
                    isSorted = false;
                    buffer = array[j];
                
                    array[j] = array[j+1];
                    array[j+1] = buffer;
                }
            }
        }
    }

    public int toIndex(int index) {
        return this.array[index];
    }

    public boolean equals(List other) {
        if ( current != other.size() ) {
            return false;
        }

        for ( int i = 0; i < current; i++ ) {
            if ( array[i] != other.array[i] ) {
                return false;
            }
        }

        return true;
    }
}
